src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -674,6 +675,76 @@ ss::future<init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
       });
 }
 
+namespace {
+
+bool is_epoch_overflow(int16_t epoch) {
+    return epoch >= std::numeric_limits<int16_t>::max();
+}
+
+bool is_valid_expected_pid(
+  const tm_transaction& tx, const model::producer_identity& expected_pid) {
+    if (expected_pid == model::unknow_pid) {
+        return true;
+    }
+
+    return tx.pid.id == expected_pid.id || ((tx.last_pid.id == expected_pid.id) && is_epoch_overflow(expected_pid.epoch));
+}
+
+bool is_valid_overflow(
+  const tm_transaction& tx, model::producer_identity expected_pid) {
+    if (!is_epoch_overflow(tx.pid.epoch)) {
+        return false;
+    }
+
+    if (expected_pid == model::unknow_pid) {
+        return true;
+    }
+
+    return tx.pid.epoch == expected_pid.epoch;
+}
+
+using current_and_last_epoch
+  = std::pair<model::producer_epoch, model::producer_epoch>;
+
+checked<current_and_last_epoch> increment_epoch(
+  const tm_transaction& tx, model::producer_identity expected_pid) {
+    current_and_last_epoch ans{-1, -1};
+
+    model::producer_epoch bumped_epoch{tx.pid.epoch + 1};
+    if (expected_pid == model::unknow_pid) {
+        // If no expected epoch was provided by the producer, bump the current
+        // epoch and set the last epoch to -1
+        ans.first = bumped_epoch;
+        ans.second = model::producer_epoch(-1);
+        return ans;
+    }
+
+    if (tx.pid.epoch == expected_pid.epoch) {
+        // If the expected epoch matches the current epoch, the producer is
+        // attempting to continue after an error and no other producer has been
+        // initialized. Bump the current and last epochs.
+        ans.first = bumped_epoch;
+        ans.second = tx.pid.get_epoch();
+        return ans;
+    }
+
+    if (tx.last_pid.epoch == expected_pid.epoch) {
+        // If the expected epoch matches the previous epoch, it is a retry of a
```

Comment:
Curious what happens if there are multiple retries. After first retry (which takes this branch) pid will be same as last_pid and the subsequent retry fails all the branches here and returns {-1, -1} ?

Reply:
Why next retries should return {-1, -1}?